### PR TITLE
add guide for udp writers

### LINKED ISSUE #23: Document procedure how to fully onboard an algorithm
Document optional/recommended/required steps to take for an interested party to register an algorithm.
Including but not limited to

- [x]  recommended way to generate a UDP representation
- [x] OGC-record for algorithm
- [x]    how to define benchmark scenario(s)
- [ ]    how to produce/include reference data
- [x] how to check validity of JSON resources

### PR DESCRIPTION
**Pull request by jdries**

### CODE REVIEW

File: udp_writer_guide.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+If you pass on the `spatial_extent' parameter to all `load_collection` processes in your UDP, then it is also allowed to
+perform filtering using vector data. This conveniently allows for advanced spatial filtering use cases!
+
+#### Temporal filtering
+
+For UDP's that support arbitrary temporal ranges, the `temporal_extent` parameter from `load_collection` is the recommended option.
+
+Many cases also require a time range with a fixed length. In such a case, you can allow to specify only the start or end
+date, and use the `date_shift` process to construct the second date in the temporal interval, ensuring a fixed length. This
```
Comment (HansVRP) [inline udp_writer_guide.md:42]: this might be already a bit more complex for entry users?

Comment (HansVRP) [reply udp_writer_guide.md:42]: (the use of date_shift)

File: udp_writer_guide.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+avoids faulty user input.
+
+
+### Avoid use of `save_result`
+
+Most openEO process graphs end in a `save_result`process. However, this is not recommended for UDPs, as the user may
+want to perform additional processing steps before generating the result.
```
Comment (HansVRP) [inline udp_writer_guide.md:49]: You can always add; something among the lines. In this way both the input as the output of your UDP will be a datacube.

File: udp_writer_guide.md [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+The description of your UDP should be quite extensive if you want users to be able to easily assess if it's suitable for them.
+We recommend including these sections:
+
+1. **Description**: A short description of what the algorithm does.
+2. **Performance characteristics**: Information on the computational efficiency of the algorithm. Include a relative cost if available.
+3. **Examples**: One or more examples of the algorithm in action, using images or plots to show a typical result. Point to a STAC metadata file with an actual output asset to give users full insight into what is generated.
+4. **Literature references**: If your algorithm is based on scientific literature, provide references to the relevant publications.
+5. **Known limitations**: Any known limitations of the algorithm, such as the type of data it works best with, or the size of the area it can process efficiently.
+6. **Known artifacts**: Use images and plots to clearly show any known artifacts that the algorithm may produce. This helps users to understand what to expect from the algorithm.
+
```
Comment (HansVRP) [inline udp_writer_guide.md:63]: you could link the example of the max NDVI here?

**Review by HansVRP — COMMENTED** [review]
minor comments, looks good.

File: udp_writer_guide.md (hunk 1 above)
Comment (jdries) [reply udp_writer_guide.md:42]: The target audience here is people trying to produce a high quality UDP within the context of an ESA project.
So we should not necessarily limit it to 'easy' things.
We can however add some examples as soon as I have any. That would probably go in a separate PR.

Comment (HansVRP) [reply udp_writer_guide.md:42]: then I would at least add a link here to the actual data_shift process

File: udp_writer_guide.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+### Parameter conventions
+
+#### Spatial filtering
+
+Any algorithm requires spatial filtering, so we can make the life of our users easier if we all use the same naming.
+
+In other openEO processes, the spatial filtering is called `spatial_extent`, and we recommend to stick to that name as 
+long as you don't have more specific names.
+
+If you pass on the `spatial_extent' parameter to all `load_collection` processes in your UDP, then it is also allowed to
```
Comment (soxofaan) [inline udp_writer_guide.md:34]:
```suggestion
If you pass on the `spatial_extent` parameter to all `load_collection` processes in your UDP, then it is also allowed to
```

File: udp_writer_guide.md (hunk 3 above)
Comment (soxofaan) [reply udp_writer_guide.md:63]: or create a template markdown file?

File: udp_writer_guide.md [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+User defined processes (UDPs) are one out of two standardized options that APEx offers to publish algorithms as a service.
+This guide gives some concrete steps and guidelines to ensure that your UDP works well for your users. These guidelines are
+written with APEx in mind, but can also serve as a general guide for openEO UDPs. Where needed, recommendations & choices 
+are made to increase uniformity across the APEx algorithm catalog.
+
+
+### Organizing your code
+
+A UDP is simply a parametrized version of an openEO workflow, and as such we recommend that you use the same code to develop
```
Comment (Patrick1G) [inline udp_writer_guide.md:11]: "workflow or process graph" - good to align these concepts

File: udp_writer_guide.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+want to perform additional processing steps before generating the result.
+
+
+### UDP documentation
+
+The description of your UDP should be quite extensive if you want users to be able to easily assess if it's suitable for them.
+We recommend including these sections:
+
+1. **Description**: A short description of what the algorithm does.
+2. **Performance characteristics**: Information on the computational efficiency of the algorithm. Include a relative cost if available.
```
Comment (Patrick1G) [inline udp_writer_guide.md:58]: hard for non platform operators to determine. What about max AOI and TOI?

File: udp_writer_guide.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+
+### Avoid use of `save_result`
+
+Most openEO process graphs end in a `save_result`process. However, this is not recommended for UDPs, as the user may
+want to perform additional processing steps before generating the result.
+
+
+### UDP documentation
+
```
Comment (Patrick1G) [inline udp_writer_guide.md:53]: maybe also add URLs to existing documentation?

File: udp_writer_guide.md (hunk 3 above)
Comment (jdries) [reply udp_writer_guide.md:70]: template added

File: udp_writer_guide.md (hunk 7 above)
Comment (jdries) [reply udp_writer_guide.md:64]: I envisioned to use the 'literature references' subsection for that.

File: udp_writer_guide.md (hunk 6 above)
Comment (jdries) [reply udp_writer_guide.md:69]: that would go under the 'known limitations' subsection I think

**Review by JanssenBrm — CHANGES_REQUESTED** [review]

File: udp_writer_guide.qmd [hunk 8]
```diff
@@ -0,0 +1,266 @@
+## Guide for writing openEO User Defined Processes {#sec-udp-writing}
```
[the remark this answers is not in the document]
Comment (jdries) [reply udp_writer_guide.qmd:1]: ok, that's logged separately, so resolving this one.

File: udp_writer_guide.qmd [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+In other openEO processes, the spatial filtering is called `spatial_extent`, and we recommend to stick to that name as 
+long as you don't have more specific names.
+
+If you pass on the `spatial_extent` parameter to all `load_collection` processes in your UDP, then it is also allowed to
+perform filtering using vector data. This conveniently allows for advanced spatial filtering use cases!
+
+#### Temporal filtering
+
+For UDP's that support arbitrary temporal ranges, the `temporal_extent` parameter from `load_collection` is the recommended option.
```
[the remark this answers is not in the document]
Comment (jdries) [reply udp_writer_guide.qmd:52]: understanding is correct, new version will be updated with your proposal.

**Comment by jdries** [comment]
latest comments have been integrated. Unless if there are major objections, I will merge on monday. We want to start creating udp's, so this guide is needed, and I assume the exposure will require further updates anyway.

**Pull request merged**